src/components/Screen.tsx

```diff
@@ -14,23 +14,39 @@ import {
 } from '../core';
 
 // Native components
-import ScreenNativeComponent from '../fabric/ScreenNativeComponent';
-import ModalScreenNativeComponent from '../fabric/ModalScreenNativeComponent';
+import NativeScreen from '../fabric/ScreenNativeComponent';
+import ModalScreenNative from '../fabric/ModalScreenNativeComponent';
 import { usePrevious } from './helpers/usePrevious';
 
-type NativeScreenProps = Omit<
-  ScreenProps,
-  'sheetInitialDetentIndex' | 'sheetLargestUndimmedDetentIndex'
-> & {
-  sheetInitialDetent: number;
-  sheetLargestUndimmedDetent: number;
+/**
+ * This messy override is to conform NativeProps used by codegen and
+ * out Public API. To see reasoning go to this PR:
+ * https://github.com/software-mansion/react-native-screens/pull/2423#discussion_r1810616995
+ */
+type SharedOverride = {
+  onAppear?: ScreenProps['onAppear'];
+  onDisappear?: ScreenProps['onDisappear'];
+  onWillAppear?: ScreenProps['onWillAppear'];
+  onWillDisappear?: ScreenProps['onWillDisappear'];
 };
 
-const NativeScreen: React.ComponentType<NativeScreenProps> =
-  ScreenNativeComponent as React.ComponentType<NativeScreenProps>;
-const AnimatedNativeScreen = Animated.createAnimatedComponent(NativeScreen);
+type NativeScreenOverrideProps = Omit<
+  React.ComponentPropsWithRef<typeof NativeScreen>,
+  'onAppear' | 'onDisappear' | 'onWillAppear' | 'onWillDisappear'
+> &
+  SharedOverride;
+
+type NativeModalScreenOverrideProps = Omit<
+  React.ComponentPropsWithRef<typeof ModalScreenNative>,
+  'onAppear' | 'onDisappear' | 'onWillAppear' | 'onWillDisappear'
+> &
+  SharedOverride;
+
+const AnimatedNativeScreen = Animated.createAnimatedComponent(
```

Comment:
It may look like I replaced one casting with another and I'm happy with separating public and native props. Still, when you look closely previously we casted NativeComponent to Component with ScreenProps, and now we leave NativeProps mostly unchanged - with clearly highlighted override.